### PR TITLE
Updated cron jobs to use git merging rather than GitHub merging

### PR DESCRIPTION
**Pull request by jmkulwik**
This updates our cron jobs to use git's merge algorithm rather than GitHub's merge algorithm when performing DTL and DTS operations automatically during the night.

We expect that this will significantly improve the success rate for these automatic operations.

While we don't have documentation directly from GitHub on their merge algorithm, it appears that within the developer community, there is consensus that the two use different merge algorithms and that git's will be significantly more likely to be successful. [Details](https://github.community/t/merge-conflicts-on-github-not-locally/949/2): 

_"GitHub compares branches from the point where they are created whereas Git in the command line generally compares branches from where they are currently. Therefore, if you have a master and a develop branch and you create a Pull Request at a certain point in time to merge develop into master but then master continues to evolve in a way that resolves a conflict before it would be created, you can get a merge conflict warning in the web interface when you wouldn’t via the command line."_

[Slack thread #1 with original discussion](https://codedotorg.slack.com/archives/C0T0PNTM3/p1648658550420619)
[Slack thread #2 with follow-up discussion](https://codedotorg.slack.com/archives/C03CK49G9/p1648834064049499)
Meeting notes: https://docs.google.com/document/d/1c7_y-KJIznTtAwRRbPPsqm6Lu-bZnkTPFxuRSFCoUZQ/edit

Next steps:
1. On the staging environment, run the following: `git worktree add ../deploy-management-repo`
2. On the staging environment, run the `merge_lb_to_staging` and `deploy_to_levelbuilder` commands as updated in this PR and ensure that they work as expected

Then as follow-up work

3. Update the DTS and DTL commands from the DOTD script to use the same logic as is updated here.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by davidsbailey** [comment]
Thanks for implementing this, Jessie! If you have not done so already, I would recommend trying this out locally, including running `bin/cron/merge_lb_to_staging`, perhaps after confirming that there isn't any content in origin/levelbuilder which would get scooped. when you do, please make sure your main repo (not the worktree) is checked out on the staging branch. one thing to know about worktrees is they can never be on the same branch as the main repo.

**Comment by dmcavoy** [comment]
Thank you for doing this Jessie. I'm going to defer to Dave and Suresh on the review of this PR as this is not an area I'm very familiar with.

**Review by davidsbailey — COMMENTED** [review]
please let me know when you are ready for another look.

**Comment by davidsbailey** [comment]
@jmkulwik I saw your note that you are planning to merge this today. did you try it out with your main repo set to the staging branch? I wasn't sure if that would cause a problem when you try to check out the same branch in your worktree.

**Comment by jmkulwik** [comment]
> @jmkulwik I saw your note that you are planning to merge this today. did you try it out with your main repo set to the staging branch? I wasn't sure if that would cause a problem when you try to check out the same branch in your worktree.

I didn't run the script itself as that would kickoff a DTS/DTL and I haven't had a chance to do that during work hours after the content editors are mostly done for the day. But I did test out the GitHub portion running just the branching logic and that worked fine. git worktree recognizes origin/staging and staging as two different branches. So long as we use `origin/<branchname>` for the base branches, we should be ok. Using `origin/<branchname>` is also safer for us since it checks out the branch in "headless" mode, and we can't push any changes from there.

File: cookbooks/cdo-apps/templates/default/crontab.erb
Comment (cat5inthecradle) [inline cookbooks/cdo-apps/templates/default/crontab.erb:69]: The `if test` and `if levelbuilder` conditionals are no longer wrapping these commands. Will these run in every environment now?

Comment (jmkulwik) [reply cookbooks/cdo-apps/templates/default/crontab.erb:69]: Not sure if I did this right, but they should be wrapped by the `if staging` conditional now.

**Comment by jmkulwik** [comment]
I tested this out from my local environment. There were a couple updates, but nothing too egregious. Full details here: https://codedotorg.slack.com/archives/C0T0PNTM3/p1650049093847099

File: lib/cdo/github.rb
Comment (davidsbailey) [inline lib/cdo/github.rb:56]: thoughts on doing something like this instead of repeatedly changing into the same directory? https://github.com/code-dot-org/code-dot-org/blob/0f5d8651842873d795a7949a746a56c2c05c9323/lib/rake/test.rake#L103

**Review by davidsbailey — APPROVED** [review]
awesome work Jessie! 

one top-level comment / question is what the error path will look like for the DOTD -- it doesn't look like any of the merge conflicts will be surfaced, so how will you recommend that one proceeds if the DTS or DTL fails?

it would be great to have approval on this from the infra side too.

File: lib/cdo/github.rb
Comment (jmkulwik) [reply lib/cdo/github.rb:56]: Oh that would be way cleaner! Thank you!! I only get a couple hours of dev time each week. If you don't mind, I'd prefer to do that as a follow-up task rather than delaying this whole change by another week or two.

**Comment by jmkulwik** [comment]
> one top-level comment / question is what the error path will look like for the DOTD -- it doesn't look like any of the merge conflicts will be surfaced, so how will you recommend that one proceeds if the DTS or DTL fails?

Hmm @davidsbailey, I'd probably recommend the DOTD follows the steps currently listed in the DOTD doc. In the past, that's how I generally see the merge conflicts get surfaced. How do the merge conflicts usually get surfaced?

**Comment by davidsbailey** [comment]
> > one top-level comment / question is what the error path will look like for the DOTD -- it doesn't look like any of the merge conflicts will be surfaced, so how will you recommend that one proceeds if the DTS or DTL fails?
> 
> Hmm @davidsbailey, I'd probably recommend the DOTD follows the steps currently listed in the DOTD doc. In the past, that's how I generally see the merge conflicts get surfaced. How do the merge conflicts usually get surfaced?

currently, the DOTD doc says for failed DTL to look at the pending PR. but it seems like there won't be one with the new system, because we abort before pushing a branch if there's a merge conflict. does that sound accurate?

File: cookbooks/cdo-apps/templates/default/crontab.erb
Comment (cat5inthecradle) [reply cookbooks/cdo-apps/templates/default/crontab.erb:69]: I don't see where that is.

Comment (davidsbailey) [reply cookbooks/cdo-apps/templates/default/crontab.erb:69]: line 50

**Comment by jmkulwik** [comment]
> > > one top-level comment / question is what the error path will look like for the DOTD -- it doesn't look like any of the merge conflicts will be surfaced, so how will you recommend that one proceeds if the DTS or DTL fails?
> > 
> > 
> > Hmm @davidsbailey, I'd probably recommend the DOTD follows the steps currently listed in the DOTD doc. In the past, that's how I generally see the merge conflicts get surfaced. How do the merge conflicts usually get surfaced?
> 
> currently, the DOTD doc says for failed DTL to look at the pending PR. but it seems like there won't be one with the new system, because we abort before pushing a branch if there's a merge conflict. does that sound accurate?

Oh oops. Apparently I always skim over that part. I'm open to suggestions. I would probably recommend that for the DTS, the DOTD go ahead and attempt the merge locally and manually resolve conflicts (following the code in the doc). For the DTL, the code would have to be adjusted slightly to use the -b option to create a new branch, but otherwise would be the same. That being said, would it be useful to find a new option for "check the PR in github?" I was assuming that everyone has their own method of resolving merge conflicts, but I'm realizing that method for some folks may be exactly what's in the doc.

**Comment by davidsbailey** [comment]
> Oh oops. Apparently I always skim over that part. I'm open to suggestions. I would probably recommend that for the DTS, the DOTD go ahead and attempt the merge locally and manually resolve conflicts (following the code in the doc). For the DTL, the code would have to be adjusted slightly to use the -b option to create a new branch, but otherwise would be the same. That being said, would it be useful to find a new option for "check the PR in github?" I was assuming that everyone has their own method of resolving merge conflicts, but I'm realizing that method for some folks may be exactly what's in the doc.

that makes sense -- the instructions to resolve the DTL issue would become a lot more like the instructions to resolve a DTS issue.

I think the main alternative that comes to mind here is to push the branch, with conflict, and still create the PR, so that the DOTD can see what the problem is, pull it down, and merge changes into it more easily (like how the DTL instructions say today). but I think starting with what you have plus DTS-like instructions is a fine start.

File: bin/cron/deploy_to_levelbuilder
Comment (sureshc) [inline bin/cron/deploy_to_levelbuilder:7]: We should schedule some follow up work to automate the configuration of the separate git worktree on staging. This would be accomplished with a Chef cookbook, and I’m hesitant to delay this change.

Comment (sureshc) [reply bin/cron/deploy_to_levelbuilder:7]: Thinking about this more … I hesitate to even create a Jira Issue to automate that because our future with Chef is uncertain. Might be useful to augment the existing comment with a `TODO`.

**Pull request merged**
